plugin/evm/validators_transitioner_test.go

```diff
@@ -0,0 +1,161 @@
+
+		// calls is expected to increase by 1 only because the cache would be used
+		got, err = validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 11, calls)
+
+		mock.ByEpochFunc = func(epoch uint64) (map[ids.ShortID]uint64, error) {
+			calls++
+			return providers, nil
+		}
+
+		got, err = validatorsTransitioner.ByEpoch(testEpoch + 1)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 13, calls)
+	})
+
+	t.Run("nominal case 2: non-nil default validators to reach the end of the function", func(t *testing.T) {
+		t.Parallel()
+
+		testEpoch := uint64(10)
+		providers := fakeProviders(testEpoch)
+		// here we give non-nil validators so that the later part of the ByEpoch() function can be reached by control and therefore tested
+		validators := fakeValidators(testEpoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				return providers, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(validators, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+
+		var totalWeight uint64
+		for _, u := range got {
+			totalWeight += u
+		}
+		assert.Equal(t, uint64(45), totalWeight)
+	})
+
+	t.Run("epoch less than 1", func(t *testing.T) {
+		t.Parallel()
+
+		testEpoch := uint64(0)
+		validators := fakeProviders(testEpoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				assert.Equal(t, testEpoch, epoch)
+				return nil, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(validators, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, validators, got)
+	})
+
+	t.Run("case for non nil error", func(t *testing.T) {
+		t.Parallel()
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				return nil, fmt.Errorf("error in getting valirators by epoch: %d", epoch)
+			},
+		}
+		validatorsTransitioner := NewValidatorsTransitioner(nil, mock)
+		providers, err := validatorsTransitioner.ByEpoch(1)
+		require.Error(t, err)
+		assert.Nil(t, providers)
+	})
+
+	t.Run("case for nil default validators", func(t *testing.T) {
+		t.Parallel()
+
+		epoch := uint64(10)
+		providers := fakeProviders(epoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				return providers, nil
+			},
+		}
+		validatorsTransitioner := NewValidatorsTransitioner(nil, mock)
+		got, err := validatorsTransitioner.ByEpoch(epoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+	})
+}
+
+func fakeProviders(epoch uint64) map[ids.ShortID]uint64 {
+
+	providers := make(map[ids.ShortID]uint64)
+
+	for i := 0; i < int(epoch); i++ {
+		providers[ids.ShortID{byte(i)}] = uint64(i)
+	}
```

Comment:
```suggestion
	providers := make(map[ids.ShortID]uint64)
	for i := 0; i < int(epoch); i++ {
		providers[ids.ShortID{byte(i)}] = uint64(i)
	}
```

When we define a variable and "fill" it in a loop, the definition should cuddle with the loop.  See [WSL Rules](https://github.com/bombsimon/wsl/blob/master/doc/rules.md) for some interesting thoughts on that. Some of them we (and specifically @awfm9) disagree with and do not follow, though, so to be read while taking it with a grain of salt 😄